### ISSUE TITLE
Use single MuseScore instance for all app windows

### ISSUE THREAD

**Issue by bkunda** [musescore/MuseScore#12647]
**Task description**

This task proposes a revision to the way MuseScore currently handles multiple open projects in MacOS. 

Currently, a new app icon appears in the dock for each opened project window. 

This is generally inconsistent with other MacOS apps that also use separate windows for each opened file, and generates both confusing UX and a cluttered UI in the dock.

Please see this design spec for an illustration of how an alternate approach might work for MuseScore 4:  
[Design-Spec_New-Project-Windows_2Aug22.pdf](https://github.com/musescore/MuseScore/files/9243039/Design-Spec_New-Project-Windows_2Aug22.pdf)

Let me know if there are any questions or comments!
Bradley

**Comment by cbjeukendrup** [comment]
The multiple app icons are forced by the multi-instances thing. One instance of MuseScore can currently open only one file at a time, and almost everything relies on the assumption that there will always be zero or one files open at a time. 

It is not possible to have multiple instances of an app but show only one app icon. Running multiple instances of one application is totally against the principles of macOS and thus not supported by Apple, and thus it affects the UX quite badly (multiple Dock icons, multiple icons in the app switcher, the shortcut to switch between windows not working...). 

So, as far as I can see, the only way to fix this, is to rewrite significant portions of code so that the app can also handle multiple files within the same instance. (For Windows, we may still need the inter-process communication system, because on Windows it is apparently much more common that multiple app windows live in multiple processes.) I don't think it's realistic that this will be fixed before the release of 4.0.

This is the reason that I've always been silently frustrated by the whole multi-instances thing, but because so far I seemed to be the only one who cares about it, I didn't make a lot of noise about it.

**Comment by RobFog** [comment]
> This is the reason that I've always been silently frustrated by the whole multi-instances thing, but because so far I seemed to be the only one who cares about it, I didn't make a lot of noise about it.

I haven't tried MuseScore 4, yet, but this is something I dread.

**Comment by bkunda** [comment]
Aha, so @cbjeukendrup is the reason that apps, like Logic and Pages in macOS, succeed in having only one icon in the dock (with multiple files open simultaneously) because they are running "multiple files within the same instance", as opposed to "multiple instances of the one application"?

FWIW this design was actually based on the UX of these two Apple apps, but I wasn't aware that something more fundamental under the hood was the reason our app behaved differently. 

If it does require a significant rewrite of code, then I worry this may need to be bumped to 4.x. ☹️

**Comment by shoogle** [comment]
I wasn't aware that "single instance, multiple windows" was a possibility. Perhaps it only works on macOS? Anyway, I think the current multi-instance approach was adopted partly because the new playback system wouldn't work otherwise, or it would require constantly loading and unloaded samples as you switch between scores. I'll be curious to see how that particular problem will be handled if we go back to using a single instance.

**Comment by cbjeukendrup** [comment]
I thought "single instance, multiple windows" is a very normal thing and as far as I know, it can work on any OS (and should even be doable with QML). Anyway, on macOS, using a single instance is "the only correct way" to do it. 

I've also always heard that the main reason for using multiple instances is something with playback, but I don't know anything more specific than that. Actually, at a first glance, the playback system seems already capable of handling multiple scores, because it works with independent "track sequences". 

I think the real difficulty will be UI state handling; some state needs to be handled within individual windows, other things need to be coordinated app-wide. For example, palettes and toolbars would need to be synced between windows. I estimate that it is not _really_ difficult to do, but just requires time to get it right, and it is guaranteed to introduce new bugs. Therefore my feeling is to delay this to 4.x, not just post beta 1. It's regrettable if it causes macOS users to be unhappy with 4.0, but even more delay + a lot of bugs is still more regrettable.

**Comment by jeetee** [comment]
By now I've seen enough remarks/complaints from users to say that the change in behavior has been voiced for the other platforms as well.

**Comment by shoogle** [comment]
@jeetee, is it just that users miss having tabs for different scores in the same app window, or does the technicality of having multiple instances cause problems on other platforms besides macOS? (What are those problems?)

**Comment by jeetee** [comment]
It's (for the other OSes) definitely all about the single windows vs multi-window. I don't think most users out there are even aware of the difference between multi-instance-single-window-single-score and single-instance-multi-window-single-score.

It was more of a statement that (since topic title here mentions MacOS) doing this with the MacOS issues as driving factor will help alleviate some of the complaints about single-vs-multi-*window* (not necessarily instance, apologies) on the other systems.

**Comment by shoogle** [comment]
Ok right. To be clear, even if we fix the multi-instance problem, it's not certain that we will go back to using a single window with tabs for every opened score. That's more of a design decision rather than a technical problem like this one, so it ought to have its own issue really.

**Comment by stcinder** [comment]
I couldn't easily find a reference in the Human Interface guidelines, although [CNET mentions it](https://www.cnet.com/tech/computing/how-to-open-multiple-instances-of-an-application-in-os-x/): On a Mac you never have multiple instances of an app. One instance runs all the windows.

FWIW, the multiple tabs in Musescore 3 is a PITA. The normal Mac way is to have multiple windows. The system will handle you putting them in tabs in a single window automatically.

**Comment by codemiester** [comment]
Personally, the tabs have been useful for doing transposition work and keeping charts in order.  I don't know how much work it would take to implement, but IMHO it should be an option to open scores in either a new window or a new tab, depending on what workflow works best for the composer/arranger.  That said, even a fix on the multiple instances would be a massive help.

**Comment by shoogle** [comment]
If you'd like to see tabs for score projects back then please open a new issue for it. This issue is about multi-instances.

**Comment by stcinder** [comment]
Tabs come for free with the MacOS, so if you can fix the multi-instance problem, it will fix the tabs at the same time.

**Comment by cbjeukendrup** [comment]
> Tabs come for free with the MacOS, so if you can fix the multi-instance problem, it will fix the tabs at the same time.

Not in a Qt app, which MuseScore is. At least not in a sensible way. It will look very awful and work badly, because macOS doesn't know about the design and color scheme of MuseScore, and MuseScore can't know when the user presses the "new tab" button that macOS generates, and can't know what to do when the user presses it.

**Comment by VictorEijkhout** [comment]
I don't care if it's tabs or windows: the multiple icons in the Cmd-TAB switcher is unlike every other application. That really needs to go. It feels like a bug, not a feature. Or maybe: an implementation detail that should not concern the user suddenly becoming visible to the user.

**Comment by GrayCatMusic** [comment]
I'm curious if this is going to be addressed soon? I'm still using MS3 mainly for this reason on my Mac.

**Comment by cbjeukendrup** [comment]
Not in 4.1. Hopefully in 4.2.

**Comment by GrayCatMusic** [comment]
That's too bad it's not sooner. I appreciate the update though. Thanks!

**Comment by MarcSabatella** [comment]
I recommend asking for help on the Support forum on musescore.org if you're having trouble figuring out how to work with multiple instances.  Although it's not ideal on macOS, it's not really that difficult either.  People on the Support site will be happy to assist.

**Comment by GrayCatMusic** [comment]
Is it a problem to ask about the status of an issue that's affecting me?

**Comment by MarcSabatella** [comment]
No, I just mean, I'd like to help you take advantage MU4 now, and so I'm suggesting you ask for help in the appropriate place for that, because the developers want to keep this issue tracker focused on the technical discussion of the issue itself, not on support.  Which makes sense.

**Comment by jeetee** [comment]
I've done some reading into how Chrome handles it's processes, and while we have some different challenges (due to Qt only having one renderThread per QApplication), I believe we could attempt a similar approach in having one Launcher process (with the ipc server) and have every other (GUI)-instance be raised as a child process into it.

Working on a small proof of concept application (without the full burden of porting and refactoring the mu::ipc usage) which hopefully might give us the insight in how to alleviate this Mac annoyance.

**Comment by jeetee** [comment]
See https://github.com/jeetee/Qt_ProcessTree_ipc for the launcher process test project. I'm not too optimistic, as on windows the child processes all do show up as separate top level entries in the task manager, but you never know..

@cbjeukendrup would be nice if you can have a local test build going on somewhere during this or next week. At best, we'll have a fairly limited refactor to fix the dock effect; at worst, we'll at least be aware that a more profound refactoring is necessary.

**Comment by cbjeukendrup** [comment]
@jeetee Thanks very much for your efforts! But unfortunately, I have bad news: it still produces multiple Dock icons on macOS. 
So I think that needs to be done is that there is always only one QApplication instance, and all QWindows live in the same process as that instance. As long as we do that, we can apply whatever sub-process magic we want, with separate audio rendering processes for example... (like Dorico, by the way)
But I'm afraid the conclusion is: a complex refactor will have to be done, regardless of whether we just try to do everything in the same process or use separate processes for audio rendering.

**Comment by frfancha** [comment]
> See https://github.com/jeetee/Qt_ProcessTree_ipc for the launcher process test project. I'm not too optimistic, as on windows the child processes all do show up as separate top level entries in the task manager, but you never know..
> 
> @cbjeukendrup would be nice if you can have a local test build going on somewhere during this or next week. At best, we'll have a fairly limited refactor to fix the dock effect; at worst, we'll at least be aware that a more profound refactoring is necessary.

The Chrome way is the right way : users are free to use tabs or different windows, and a tab can be simply dragged and dropped in its window. That's the best of both world.
I wouldn't worry about each tab appearing as its own process in task manager, that's expected.

**Comment by shoogle** [comment]
I think Chrome deliberately uses separate processes so that if one tab crashes it doesn't bring down the entire program (at least that's how it works in theory...). But it sounds like that approach wouldn't work for us anyway.

**Comment by frfancha** [comment]
> I think Chrome deliberately uses separate processes so that if one tab crashes it doesn't bring down the entire program (at least that's how it works in theory...). But it sounds like that approach wouldn't work for us anyway.

Yes that's true, but how wouldn't that work for you? I thought having separate processes was a req for you to manage different sounds? It is a genuine question I'm a IT dev and would like to help if something is doable with my capacity.

**Comment by jeetee** [comment]
> I wouldn't worry about each tab appearing as its own process in task manager, that's expected.

It's not about the process appearing, it's about them not appearing as children (even when they are). Take note that if you look at the current MuseScore process, you'll already notice that the crashpad handler exists as it's child process (and is listed as such).

The difficulty here (and why casper and myself were/are testing out some approaches) is to make Q(Gui)Application instances behave according to the Chrome model. If you'll run through the linked test project, you'll notice when running/debugging it already is one top level process with everything else as a child (or child of child) under it. However, due to how QGuiApplications register themselves with the different OSses, each of the initial child processes is still regarded as "top level" by windows (task manager) and MacOS (multiple dock icons, instead of one).

The additional challenge posed (for the "big refactor", which seems a more likely requirement for every other failed path attempt) is that Qt only allows for a single UI renderThread for a topLevel QApplication. This would mean that we'd probably need to come up with an implementation to support multiple renderThreads to re-enable a multidocument interface without making it slower than the current MuseScore implementation.
Most likely this means a change to how the ScoreView is implemented to allow it to do buffered rendering instead of inside an existing QML component and then transfer that buffer via Shared Memory to the actual UI renderThread.

Another question that is yet to be answered, is whether the cause of the multiple top-level entries is the presence of multiple Q(Gui)Applications (more likely) or the fact that they each spawn their own QMainWindow (less likely). Though still allowing multiple QMainWindows in a single multi-processed QGuiApplication might still mess with MacOS's unified menubar behavior.
If it does, then aside from ScoreView, the normal MainWindow rendering and how palettes/mixer etc are drawn within it will also require an architectural review.

**Comment by jeetee** [comment]
I think the possibilities research can now be pretty much concluded after diving into [some](https://forum.qt.io/topic/66773/run-parts-of-an-application-in-separate-process/7) [more](https://forum.qt.io/topic/151769/run-qt-gui-in-separate-thread) [topics](https://www.kdab.com/the-eight-rules-of-multithreaded-qt/) around this; with a fairly decent summary in [this topic around PyQt and multithreading](https://forum.qt.io/topic/124878/running-qapplication-exec-from-another-thread-qcoreapplication-qguiapplication). I'll summarize the most important takeaways relating to MuseScore here:

1. A process can have only one QCoreApplication or derivative (QApplication/QGuiApplication) of it.

2. The QCoreApplication object **must** be the 1st QObject created on the "main thread".
The "main thread" according to Qt is the first thread that creates a QObject (even if it's not the QCoreApplication). This does not have to coincide with the actual application thread running the main function.
*Consequence:* You can't use a QThread to run a QCoreApplication, as QThread itself is already a QObject.

3. Due to thread affinity of QObjects, the UI **really** *should* run within the eventloop of the QCoreApplication. As such there should be only one UI thread and it will be the "main thread".

4. (At least on Windows) Spawning multiple QMainWindows is just fine and likely doesn't enforce multi-dock-behavior on MacOS.

5. The child processes showing up as top level processes in the windows task manager in the ProcessTree test project from above are indeed due to violating the "single QCoreApplication" rule. The way QCoreApplication handles registering itself to the OS enforces them as seperate toplevel processes (but with child-linkage).

6. It is possible to perform (part of) the drawing in a separate thread and doing so in shared memory if that drawing is computationally intensive (see [the Qt Mandelbrot Example](https://doc.qt.io/qt-5/qtcore-threads-mandelbrot-example.html)).
Whether or not this is something we'd want to use for ScoreView or not remains to be seen and subject of further analysis (by "someone"). It might already suffice to perform our layoutcalculations on a worker thread while leaving the drawing based on the resulting values within the UI thread.

I'll have a go at working these conclusions (aside from the renderthread approach) into a new version of the ProcessTree repo so we can confirm that his method of single-instance running will address the MacOS docking problem.

**Comment by jeetee** [comment]
Finally got around to updating the example project to be a single (gui) instance build. For which Casper could confirm the desired dock icon behavior on MacOS.

Next steps:
Figure out which modules can be loaded/linked to the Single GUI Instance (settings/workspaces come immediately to mind, avoiding the synchronization mess we currently have on them) and which modules are MainWindow-specific (notationcontroller etc).

This does not yet begin to investigate a possible tabbed multi-document single-window approach. Only looking into single instance, multi window to limit/minimize the changes needed for MuseScore.

I'll probably halt development on the separate test project unless something specific needs additional attention and start moving into a MuseScore branch for the bulk of those changes; so we can account for the MuseScore specific interactions and architecture.

**Comment by jeetee** [comment]
I'm having a (slow!) go at trying to modify MuseScore on the main/app/appshell level over at https://github.com/jeetee/MuseScore/tree/12647-single-instance-model . As always, anyone with time is free to pick this up as I can't guarantee free time.

And for those following along in hopes of seeing this fix. What I'm doing now is only the starting step; it'll enforce single instance and multi window behavior. But as many things now assume single instance usage, I'm expecting a lot of things will need further updates/fixing after the initial structural changes.

**Comment by igorkorsukov** [comment]
Two ideas.
1. I don’t have MacOS, but in Ubuntu (Gnome) applications can be grouped - one icon, several instances, and there is a menu for managing them. I don’t know how this is implemented, if you don’t do anything, several application icons appear (for example, for MuseScore), i.e. for such a grouping you need to do something. After doing a little research, I couldn't find what should be do for it..
But the idea is that if there is such an opportunity in Gnome, then maybe there is something similar in MacOS. We just need to find out what exactly needs to be done for this.

![image](https://github.com/musescore/MuseScore/assets/3818029/5ca887f0-61a6-4683-baac-f417fa204307)

2. We can do a hack. 
It is possible to show a window without an icon on the taskbar.
On the other hand, we always have a IPC server that “knows” about all instances.
Thus, we can make the icon appear only for one instance, for the IPC server, and not appear for the rest.
All that remains is to decide how to switch to other instances.
Probably we can make some kind of menu...
Or simulate their display through fake child windows, which, when selected, will actually cause the display of the corresponding instance.

**Comment by jeetee** [comment]
or 3: actually do run a single UI instance?

**Comment by igorkorsukov** [comment]
> or 3: actually do run a single UI instance?

I see several problems with this: 
* Any crash will lead to closing of all open windows without any saves
* This is more difficult to develop, we will need to either clearly understand what is common and what is different... or pass a some context everywhere
* It's strange to make these complications due to a limitation of one of the platforms

**Comment by cbjeukendrup** [comment]
@igorkorsukov Both options seem a no-go. 

1. Such thing doesn't exist on macOS. Grouping of dock icons is not possible. 
You could say each dock icon corresponds to a NSApplication instance, so there should be only one of those (and thus only one process that creates a QApplication). 

2. This hack doesn't really solve the problem. You would still see multiple icons in the App Switcher, and OS-wide shortcuts like Cmd+Tab and Cmd+` don't work. There is simply a whole stack of problems that are either unfixable or require even more ugly hacks. This is not sustainable. 

> It's strange to make these complications due to a limitation of one of the platforms

You could say just as well: when building an app for macOS, it's weird if we don't follow the correct architecture of a macOS app. Literally all macOS apps use one main process. 

There are two things that are fine:
- use background processes, if we really want things to be in separate processes. But still, have one process that owns the NSApplication and all windows. 
- continue using the multi-process architecture on other platforms, but simply _add_ the possibility to open multiple scores in one process for macOS. 

One more remark: the multiinstances architecture is very limiting. For example, "unroll repeats" can't really be implemented now: it should create a score in memory based on the current score in its current state, and open that. That's impossible with the multiinstances thing: you'd need to write the generated score to disk, then create a new instance and tell it to open that score and treat it as a special newly created score. This involves the addition of even more "private" command line options that you actually don't want to exist. 
Or what about the extensions framework, where users will expect the possibility to do something similar to "unroll repeats", or wants to perform some action for all open scores (including ones in different instances). 
Or what about the Split View feature, that lets you open two scores side by side in one window, or in multiple tabs? This seems simply impossible with the current architecture. 

Oh, and there is performance and memory usage. Opening a score doesn't take long usually, but starting up a new instance does take very long (sometimes even more than 20 seconds!). And everything is duplicated in memory for every score that's open. 

My conclusion: both options have _different_ pros and cons, but the current option is absolutely not that much better than the "proper" way. The proper way has the advantage of being ehm... _proper_, and gives more possibilities for multi-score features. 
I'm not too convinced about the "a crash would affect only one window" argument: in the first place, there should not be any crashes, and in the second place, that's what AutoSave and session restore are for. (Incidentally, the "restore session" feature is basically broken by multiinstances.) 

Yes, state management is more difficult with multiple scores in one instance. But I think that's just part of our job. I'd rather work two months on a solution that we can be proud of than two weeks on a workaround that only replaces the problem with 5 other problems.

**Comment by igorkorsukov** [comment]
I don't agree with these arguments, but I need to think and investigate ) 
Are all large applications that are not specific to MacOS, for example Photoshop, AutoCAD, MS Office, etc., running many windows in one process..?

**Comment by stcinder** [comment]
> Are all large applications that are not specific to MacOS, for example Photoshop, AutoCAD, MS Office, etc., running many windows in one process..?
> 
Every app does this on MacOS.

**Comment by igorkorsukov** [comment]
> Every app does this on MacOS.

Help me figure it out.
For example, I have some Sql Manager, it opens one database.
I want to open another base.
I click on the Sql Manager launcher and the previous instance opens with the database already open.
How can I open another database on MacOS?

**Comment by cbjeukendrup** [comment]
There are of course apps that support only one open file at a time, because the developer did not take the effort to implement multifile support.
But even the most complicated apps (Dorico, Sibelius, Logic Pro, Xcode) support multiple windows with multiple projects, all within one (main) process.

**Comment by shoogle** [comment]
Chrome has a [separate process](https://support.google.com/chrome/thread/2064254?hl=en&msgid=2065044) for each tab and each extension, plus a single process for all GPU rendering. Though there is probably still a single main process on macOS that communicates with the subprocesses somehow.

Of course Chrome is also known for being a memory hog, but other Chromium-based browsers don't seem to have that problem.

**Comment by wizofaus** [comment]
@cbjeukendrup re "For example, "unroll repeats" can't really be implemented now: it should create a score in memory based on the current score in its current state, and open that"

I actually made a good start on that and it 90% of the way there. It was working fine as an option just to "unroll in-place", but the other option was simply to save out a new score file (add -unrolled to the current name) and open that in the 2nd instance.
I don't really see how it's connected to being able to have multiple top-level windows in a single process.
The big challenge there is the use of effectively global variables via the ioc module - it would have to be smarter and choose the right IOC context depending on which window an event was associated with.

**Comment by stcinder** [comment]
I think the easiest solution would be to write the user interface components as a mac app. There is no point continuing to develop for OSX if this problem is not fixed.

**Comment by rpatters1** [comment]
While I agree that the current multi-instance implementation is counter to Apple's UX guidelines, I actually like it because you can use cmd-Tab to go between open documents. That has always been one of the major frustrations I've had with Finale (which I mention only because it is the single desktop app I have the most experience with).

I do know about tabs, but I dislike them. They add clutter and occupy real estate.

One of the features of Finale that I like quite a lot is the ability to open multiple windows on the same document. Perhaps this is significant scope-creep, but that turns out to be useful above and beyond viewing different parts. Especially in a multiple-monitor setup.

**Comment by cbjeukendrup** [comment]
Cmd+Tab is indeed the shortcut to switch between different apps (or app instances); but there is also Cmd+` (i.e. the key next to the Z key) to switch between different windows of the same app (should work with Finale too!). So switching between documents can be done just as easily when everything is in one app instance. 

Opening multiple windows of the same score is most likely outside the scope of this issue. But I agree that we should implement this. Ultimately I think we'll want to support all imaginable configurations, like having different files as different tabs in the same window, or side-by-side instead of as tabs, or different files in multiple windows, or the same file in multiple windows; or any combination. It's absolutely not realistic to implement that all at once, but we should keep that in mind when designing a MVP fix for this issue.

**Comment by rpatters1** [comment]
>Cmd+`

Holy smokes. I wish I knew about this before Finale was a corpse! 🤣 (Indeed it works in Finale.)

**Comment by mkruselj** [comment]
I noticed that I replied to the [now-defunct issue tracker](https://musescore.org/en/node/338910#comment-1273487), heh. So let me repeat what I wrote there, here.

I think this is kind of a cop-out answer that this behavior is now "by design". No, you broke a workflow that was there before.

The playback engine itself could be a separate process, acting as a server, which loads ALL the sounds that EVERY opened score project needs (and, of course, unloads sounds that are not used anymore when a score is closed). This way you don't need to invoke separate MuseScore instances, and you could revert to the nicer UX of MS3 with score tabs (way nicer IMHO). And since all required sounds would be loaded onto the playback server, switching tabs would be instantaneous. We have the RAM these days, use it!

This is also how software like Vienna Ensemble Pro works. It's basically a host for your sounds, you could open multiple DAW projects and they could all hook in to all the sounds you have in a VEPro instance.

I don't see what prevents MuseScore to work exactly the same way (apart from obviously coding the playback server up and all the ancillary things required).

**Comment by cbjeukendrup** [comment]
I don't think it was ever (officially) said that this is by design, because it isn't really. I'd rather say it's a "known issue". Given the amount of refactoring required to fix it, it seems a good candidate for MuseScore 5.0.

> I don't see what prevents MuseScore to work exactly the same way (apart from obviously coding the playback server up and all the ancillary things required).

I'm not really the best person to answer this question, but here's what I know:
- Transferring a complex data structure like a score or playback data (which is much more complicated than a simple list of note on/off events) over a process boundary in real-time is difficult. Constantly serialising everything (e.g. to JSON) and then sending it over web sockets between processes seems no option to me because it would be way too inefficient. Still, I guess VEPro must be doing something like that, because I seem to recall it can also work over the network. So they must be sending serialised VST events in the one direction, and pieces of rendered audio buffers in the other direction. Anyway, networking or inter-process communication are not my fields of expertise, so I can't say much about this.
(Also, as far as I know, the option to do everything just in one process (and load VST instances etc. for multiple scores at the same time) has not definitively been ruled out.)
- Playback is only part of the problem; something like UI state management becomes more complex when there can be more than one score open at a time. However, how complex it really is, depends very much on what we want to do exactly: if we accept to go back to single-window only and have scores in separate tabs, it's quite doable, but if we also want to support having multiple windows in one process, there's a bit more to it (and it is even uncertain whether this would be possible with the KDDockWidgets framework that we use, but perhaps that's also solvable). This part is not really difficult, but just a lot of work, and relatively bug-risky. That's why it can't be done casually between other tasks.

**Comment by mkruselj** [comment]
JSON is indeed not efficient for this sort of stuff. MessagePack, on the other hand, is a different story (I reckon this is what VEPro is using for sending event data, whereas audio buffers are simply streamed back).

**Comment by Ptujec** [comment]
Yes, please. The current behaviour is just wrong. No other app is doing this on macOS. It is cluttering up the dock. I love the direction MuseScore is going. It is so improved. Except for this. This is really annoying.

**Comment by markchagers** [comment]
On a different note: the fact that every window has one single tab seems to contradict the whole decision to switch to one document per instance: that empty tab bar is literary screaming for more tabs.
As noted by others, MacOS apps never have multiple instances open, which make MS stand out as very un-mac-like.
Others seem to like the tab-less interface better, so the natural thing to do would be to make it an optional feature (and when off, the tab bar should not be visible at all IMO).

**Comment by SteveBlower** [comment]
You get more tabs when you open parts, one tab per part.   It stops screaming then.

**Comment by mkruselj** [comment]
Then it makes sense to hide the tabbar completely until any parts are opened.

I still consider this a horrific regression in UX, though.

**Comment by markchagers** [comment]
> I don't think it was ever (officially) said that this is by design, because it isn't really. I'd rather say it's a "known issue". Given the amount of refactoring required to fix it, it seems a good candidate for MuseScore 5.0.
> 

I understand that a project like MuseScore depends on a lot of work by a lot of people, and I'm grateful for the work that's being done.
I would happily wait for this to be fixed in 5.0 as long as I know it's at least on the roadmap.

**Comment by kdekooter** [comment]
Any news on the progress of this issue? Meanwhile the UI still gives the impression that multiple score tabs are possible:

<img width="499" height="134" alt="Image" src="https://github.com/user-attachments/assets/c60206d2-3254-4150-a6c1-fa7fb27f1e6f" />

**Comment by jeetee** [comment]
> Any news on the progress of this issue?

If there was, you'd have seen the update appear right here, in this thread.

> Meanwhile the UI still gives the impression that multiple score tabs are possible

The tabs there are for the master score and part scores opened within the same file. Open any part of your score and you'll see the tab for it appear there.

**Comment by kdekooter** [comment]
@jeetee fair enough, thanks for your reply.

**Comment by erinic04** [comment]
Is this fixed for 4.7.4? last comment I see is from the PR, which says it's enabled by default in the master, but the actual thing doesn't seem to be fixed, still multiple taskbar icons

**Comment by cbjeukendrup** [comment]
The first release containing the fix will be 5.0.0. From a technical point of view, this could be considered one of the headline features, even though end users will consider it not more than a bug fix...

**Comment by erinic04** [comment]
alright cool, good to know!
